Genetic Algorithms on GPU using Intel Distribution of Python numba-dpex code sample (#2296)
oneapi-src/oneAPI-samples: AI-and-Analytics/Features-and-Functionality/IntelPython_GPU_numba-dpex_Genetic_Algorithm/README.md

If the default path is used during the installation of Intel® Distribution for Python*:
```
$HOME/intelpython3/bin/python -m ipykernel install --user --name=base
```
If a non-default path is used:
```
<custom_path>/bin/python -m ipykernel install --user --name=base
```
**2. Launch Jupyter Notebook** 
<!-- add other flags to jupyter notebook command if needed, such as port 8888 or allow-root -->
```
jupyter notebook --ip=0.0.0.0
```
**3. Follow the instructions to open the URL with the token in your browser**

**4. Select the Notebook**
<!-- add sample file name -->
```
IntelPython_GPU_numba-dpex_Genetic_Algorithm.ipynb
```
**5. Change the kernel to `base`**
  <!-- specify relevant kernel name(s), for example `pytorch` -->
**6. Run every cell in the Notebook in sequence**

## Example Output

If successful, the sample displays `[CODE_SAMPLE_COMPLETED_SUCCESSFULLY]` at the end of execution. The sample will print out the runtimes and charts of relative performance with numba-dpex and without any optimizations as the baseline. Additionally, sample will print the best and worst path found in the Traveling Salesman Problem.

## Related Samples

> * [Get Started with the Intel® Distribution for Python*](https://www.intel.com/content/www/us/en/developer/articles/technical/get-started-with-intel-distribution-for-python.html)

Reviewer: Usually in this section we provide the links to other samples from the repo. In this case, it is https://github.com/oneapi-src/oneAPI-samples/tree/master/AI-and-Analytics/Features-and-Functionality/IntelPython_Numpy_Numba_dpex_kNN. This way user will stay on GitHub instead of going to IDZ and then returning back, which makes user path easier

| What you will learn     | How to implement the genetic algorithm using the Data-parallel Extension for Numba* (numba-dpex)?
| Time to complete        | 8 minutes
| Category                | Code Optimization

>**Note**: The libraries used in this sample are available in Intel® Distribution for Python* as part of the [Intel® AI Analytics Toolkit (AI Kit)](https://software.intel.com/en-us/oneapi/ai-kit).

## Purpose

In this sample, you will create and run the general genetic algorithm and optimize it to run on GPU using Intel® Distribution for Python* numba-dpex. You will learn what are selection, crossover and mutation, and how to adjust those methods from general genetic algorithm to a specific optimization problem which is the Traveling Salesman Problem.

## Prerequisites

| Optimized for           | Description
|:---                     |:---
| OS                      | Ubuntu* 20.04
| Hardware                | GPU
| Software                | Intel® AI Analytics Toolkit (AI Kit)

### For Local Development Environments

You will need to download and install the following toolkits, tools, and components to use the sample.

- **Intel® AI Analytics Toolkit (AI Kit)**

>   You can get the AI Kit from [Intel® AI Analytics Toolkit](https://www.intel.com/content/www/us/en/developer/tools/oneapi/ai-analytics-toolkit-download.html).

Reviewer: I'm not sure if it is right to point user to AI Tools for this specific sample. IDP is not included into AI Tools, only some packages are included, and we don't mention IDP on AI Tools pages. Maybe we need to point to standalone IDP instead?

Another thing - the overall README needs DX review to ensure instructions are up-to-date. We have a README template that can help with that, but for this specific sample some adjustments may be needed. There should also be a trademarks review by InfoDev.

Let's discuss more via e-mail and we can also schedule a short meeting

#### Activate Conda

1. Activate the Conda environment.
   ```
   conda activate base
   ```
   By default, the AI Kit is installed in the `/opt/intel/oneapi` folder and requires root privileges to manage it.

   You can choose to activate Conda environment without root access. To bypass root access to manage your Conda environment, clone and activate your desired Conda environment using the following commands similar to the following.

   ```
   conda create --name usr_base --clone base
   conda activate usr_base
   ```

#### Run the Python Script

1. Change to the sample directory.
2. Run the script.
   ```
   python IntelPython_GPU_numba-dpex_Genetic_Algorithm.py
   ```

#### Run the Jupyter Notebook (Optional)

1. Launch Jupyter Notebook.
   ```
   jupyter notebook --ip=0.0.0.0
   ```
2. Follow the instructions to open the URL with the token in your browser.
3. Locate and select the Notebook.
   ```
   IntelPython_GPU_numba-dpex_Genetic_Algorithm.ipynb
   ```
4. Run every cell in the Notebook in sequence.

#### Troubleshooting

> If you receive an error message, troubleshoot the problem using the **Diagnostics Utility for Intel® oneAPI Toolkits**. The diagnostic utility provides configuration and system checks to help find missing dependencies, permissions errors, and other issues. See the [Diagnostics Utility for Intel® oneAPI Toolkits User Guide](https://www.intel.com/content/www/us/en/develop/documentation/diagnostic-utility-user-guide/top.html) for more information on using the utility.

Reviewer 2: ```suggestion
If you receive an error message, troubleshoot the problem using the **Diagnostics Utility for Intel® oneAPI Toolkits**. The diagnostic utility provides configuration and system checks to help find missing dependencies, permissions errors, and other issues. See the [Diagnostics Utility for Intel® oneAPI Toolkits User Guide](https://www.intel.com/content/www/us/en/docs/oneapi/user-guide-diagnostic-utility/current/overview.html) for more information on using the utility.
```

## Example Output

> If successful, the sample displays `[CODE_SAMPLE_COMPLETED_SUCCESSFULLY]` at the end of execution. The sample will print out the runtimes and charts of relative performance with numba-dpex and without any optimizations as the baseline. Additionally, sample will print the best and worst path found in the Traveling Salesman problem.

Reviewer 2: ```suggestion
If successful, the sample displays `[CODE_SAMPLE_COMPLETED_SUCCESSFULLY]` at the end of execution. The sample will print out the runtimes and charts of relative performance with numba-dpex and without any optimizations as the baseline. Additionally, sample will print the best and worst path found in the Traveling Salesman Problem.
```

>**Note**: Before running the following commands, make sure your Conda/Python environment is activated

```
pip install -r requirements.txt
pip install notebook
``` 
For Jupyter Notebook, refer to [Installing Jupyter](https://jupyter.org/install) for detailed installation instructions.

## Run the Sample
>**Note**: Before running the sample, make sure [Environment Setup](https://github.com/oneapi-src/oneAPI-samples/tree/master/AI-and-Analytics/Getting-Started-Samples/INC-Quantization-Sample-for-PyTorch#environment-setup) is completed.

Go to the section which corresponds to the installation method chosen in [AI Tools Selector](https://www.intel.com/content/www/us/en/developer/tools/oneapi/ai-tools-selector.html) to see relevant instructions:
* [IDP Offline Installer (Validated)](#ai-tools-offline-installer-validated)
* [Conda/PIP](#condapip) 

### AI Tools Offline Installer (Validated)

**1. Register Conda kernel to Jupyter Notebook kernel**

If the default path is used during the installation of AI Tools:
```
$HOME/intel/oneapi/intelpython/envs/<offline-conda-env-name>/bin/python -m ipykernel install --user --name=<offline-conda-env-name>
```
If a non-default path is used:
```

> <custom_path>/bin/python -m ipykernel install --user --name=<offline-conda-env-name>

Reviewer: ```suggestion
<custom_path>/bin/python -m ipykernel install --user --name=base
```

# `Genetic Algorithms on GPU using Intel® Distribution for Python numba-dpex` Sample

The `Genetic Algorithms on GPU using Intel® Distribution for Python numba-dpex` sample shows how to implement a general genetic algorithm (GA) and offload computation to a GPU using numba-dpex.

| Property                    | Description
| :---                    | :---
| Category                | Code Optimization
| What you will learn     | How to implement the genetic algorithm using the Data-parallel Extension for Numba* (numba-dpex)?
| Time to complete        | 8 minutes

> >**Note**: The libraries used in this sample are available in Intel® Distribution for Python* [Intel® Distribution for Python*](https://www.intel.com/content/www/us/en/developer/tools/oneapi/distribution-for-python.html).

Reviewer: In the template we have a note about validation here. We can adjust it to IDP the following way:

> **Note**: This sample is validated on Intel® Distribution for Python* Offline Installer and AI Tools Offline Installer. For the full list of validated platforms refer to [Platform Validation](https://github.com/oneapi-src/oneAPI-samples/tree/master?tab=readme-ov-file#platform-validation).

| Hardware                | GPU

> | Software                | Intel® AI Analytics Toolkit (AI Kit)

Reviewer: ```suggestion
| Software                | Intel® Distribution for Python*
```

## Key Implementation Details

This sample code is implemented for GPUs using Python. The sample assumes you have numba-dpex installed inside a Conda environment, similar to what is installed with the Intel® Distribution for Python*.

> >**Note**: Read *[Get Started with the Intel® Distribution for Python*](https://www.intel.com/content/www/us/en/developer/articles/technical/get-started-with-intel-distribution-for-python.html)* to find out how you can achieve performance gains through Intel optimizations.

Reviewer: Maybe we could add a link to this article into the Related Samples section? It seems logical to put it after we explained everything about this specific sample

The sample tutorial contains one Jupyter Notebook and one Python script. You can use either.

## Environment Setup
You will need to download and install the following toolkits to use the sample.
<!-- Use numbered steps instead of subheadings -->

**1. Get Intel® Distribution for Python***

If you have not already, install Intel® Distribution for Python* via [Installer](https://www.intel.com/content/www/us/en/developer/tools/oneapi/distribution-python-download.html?operatingsystem=linux&linux-install-type=offline). AI and Analytics samples are validated on AI Tools Offline Installer. It is recommended to select Offline Installer option in AI Tools Selector.

> **2. (Offline Installer) Activate the AI Tools bundle base environment**

Reviewer: For IDP the only option is offline installer, so no need to additionally specify this in brackets
```suggestion
**2. Activate the AI Tools bundle base environment**
```

This sample code is implemented for GPUs using Python. The sample assumes you have numba-dpex installed inside a Conda environment, similar to what is installed with the Intel® Distribution for Python*.

The sample tutorial contains one Jupyter Notebook and one Python script. You can use either.

> ## Environment Setup

Reviewer: step 2 line 43 - probably it should be 'intelpython3' instead of 'intelpython'

Reviewer: step 3 - this step looks redundant because we only have 1 env for IDP and user is activating it on step 2
